Add test framework for C++ codes (googletest)

yjaeseok opened the pull request:
We need a test framework for C++ codes.
So I add googletest package as submodule.
And then make some testcase for calculator

After this patch, we can try to test for C++ codes.

$ mkdir build
$ cd build
$ cmake ..
$ make
$ ./unit_tests

You can see test results like below:

[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from Calculator
[ RUN      ] Calculator.Add
[       OK ] Calculator.Add (0 ms)
[ RUN      ] Calculator.Sub
[       OK ] Calculator.Sub (0 ms)
[ RUN      ] Calculator.Mul
[       OK ] Calculator.Mul (0 ms)
[ RUN      ] Calculator.Div
[       OK ] Calculator.Div (0 ms)
[----------] 4 tests from Calculator (0 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (1 ms total)
[  PASSED  ] 4 tests.

ISSUE=#47

yjaeseok committed:
Add test framework for C++ codes (googletest)

We need a test framework for C++ codes.
So I add googletest package as submodule.
And then make some testcase for calculator

After this patch, we can try to test for C++ codes.

$ mkdir build
$ cd build
$ cmake ..
$ make
$ ./unit_tests

You can see test results like below:

[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from Calculator
[ RUN      ] Calculator.Add
[       OK ] Calculator.Add (0 ms)
[ RUN      ] Calculator.Sub
[       OK ] Calculator.Sub (0 ms)
[ RUN      ] Calculator.Mul
[       OK ] Calculator.Mul (0 ms)
[ RUN      ] Calculator.Div
[       OK ] Calculator.Div (0 ms)
[----------] 4 tests from Calculator (0 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (1 ms total)
[  PASSED  ] 4 tests.

ISSUE=#47

yjaeseok committed:
Merge branch 'master' into master

romandev reviewed (commented):
Wow! Thank you very much @yjaeseok!
(Also, please update AUTHORS file in this patch)

How can I use this test?
I don't have cmake in my environment and we are using GYP instead of cmake.
If it's not easy to integrate GYP now, could you update dependent packages to README.md for all platforms.

(Also, I think the unit tests should run when users execute ```./bacardi test``` command)

romandev commented:
Anyway very nice commit description. Your description is a best practice to me.

yjaeseok commented:
Thank you for your comments!
As you mentions, I will try how to integrate gyp module CMake, and combine `./bacardi test `

yjaeseok committed:
add cmake-js to package.json

yjaeseok committed:
Add AUTHORS jaeseok yoon

romandev commented:
Yeah, but this may be a big deal than I thought.
Therefore, I'm okay to merge this patch as-is. (including AUTHORS and minimum README.md update)
and then you can migrate cmake to GYP.
I recommend to split this large work into smaller patches. (I think)

yjaeseok commented:
Okay, anyway, which platform do you use ? Windows? Mac? Linux? else?..

romandev commented:
@yjaeseok I'm using Mac and Linux. (But we are supporting Windows as well :))

yjaeseok committed:
Update README.md and test script for C++ unit test

I update README.md files for introduce we need CMake.
But this is a temporary patch, I will update CMake to
GYP like our build system.

And User can testing `./bacardi test` command C++ unit
test also

yjaeseok commented:
I updated README.md files for indicating we need to setup CMake program for C++ unit testing.
(But this is just temporary patch as @romandev mention !! )

And If user has CMake program, user can C++ unit test `./bacardi test` as run command only

yjaeseok committed:
README.md link error fix

yjaeseok committed:
Delete cmake-js from package.json

yjaeseok committed:
modify README.md and test script

yjaeseok committed:
test script update

yjaeseok commented:
hmm.. 
This patch operate only unix/linux system.
Because of Make command. 

I will check it soon.

yjaeseok commented:
I will consider how to make `gyp` `make` command.

yjaeseok committed:
Change test command for windows platform

yjaeseok committed:
Update CMakeLists.txt, README.md, test command for unit testing

yjaeseok commented:
Finally, even though we use cmake module for buliding, I got successful unit test result in Windows Platform. 
For later, i will try to change cmake to gyp for building system and add unit test result test command also.

yjaeseok commented:
After demo, please review again @romandev , @hwanseung

yjaeseok committed:
remove unit test command in test command

yjaeseok commented:
Oh... T.T
for Linux or Unix system, target library must be changed for google test code.

So I should update CMakeLists.txt file.. OTZ ..

romandev commented:
Cheer up! It's not easy to support multi-platforms.
My PR is also blocked by Windows platform although it works well in Linux and Mac. :)

yjaeseok committed:
Update to CMakeLists.txt for supporting Linux

yjaeseok commented:
Please review. I updated to support Linux and Windows System.

romandev commented on README.md:
````diff
@@ -28,6 +29,24 @@ native layer.
 - You can just run the following command to test.
   - ```./bacardi test```
 
+## Unit Test for C++ module
````
I thought that this is necessary but my thinking was changed.
Sorry for that..

It's a WIP feature, so, I think it's better to hide this description.
until this feature is enabled fully.
WDYT?

yjaeseok replied:
Okay, I agree your comment! After this feature is enabled fully. I will re-consider update README or not.

romandev commented on bootstrap/command/test:
```diff
@@ -33,3 +33,16 @@ if is_already_formatted; then
   echo "You should run |bacardi format| command first."
 fi
+
+#if is_windows_platform; then
```
This project is not Chromium but we are following Chromium style and policy if possible.
In Chromium, we try to avoid unused code or unused comments if possible.
Once the code is merged, it requires some reasons to remove them.
So, IMHO, it's better to remove these comment lines. WDYT?

yjaeseok replied:
Ah... Thanks to you, I came to know Chromium Style. I will remove this comments!

romandev commented on test/calculator.cc:
```diff
@@ -0,0 +1,39 @@
+/*
```
I discussed about this test with @hwanseung but we thought that this test might not be necessary.
Because the calculator example is already tested by Mocha with JS.
But anyway, we need an test example for unittest as well.
So, I think we might have two options as follows:
  - Remove this test and rewrite a new test we really need (e.g. unit test for NativeTypeTraits)
  - Keep this test as-is for now.

The first option is good to me but I also don't want this PR to be pending.
So, the second option is also good to me.

WDYT? @hwanseung @yjaeseok

yjaeseok replied:
I choose to remove this test and rewrite a new test we really need 
But I am having trouble adding new test code right now.

So I would like to add only testing framework at this PR.

romandev replied:
@yjaeseok okay, it's a good idea. (and Congratulation on the birth of junior Jake!)

yjaeseok replied:
But....... If I do this, this PR becomes for .gitmodules only. T.T

yjaeseok replied:
So I will make the new test ASAP!

romandev replied:
@yjaeseok Ah, sorry for confusing.
I meant that we might not need calculator test.
Your CMakeList.txt and main() is still needed I thought.
But you can do it in second patch, which do you prefer?

yjaeseok committed:
Revert some unit test code and Make files

We decide some unit test code we do not needed.
So once remove this test and rewrite a new test
we really need.

yjaeseok commented:
@romandev I update this PR, please re-review!
(My wife is sleeping now :+1: )

romandev reviewed (approved):
lgtm
